From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sat, 6 May 2023 11:39:16 +0200
Subject: [PATCH] PackException class

Adds the `PackException` class, so that pack reading and loading can fail with a clear message and cause.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackException.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackException.java
new file mode 100644
index 0000000000000000000000000000000000000000..536e4b2ea56723c53d6f9c377198e86f7d3eb4c6
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackException.java
@@ -0,0 +1,18 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+/**
+ * A checked exception that occurred while scanning, parsing, reading or loading part of a pack.
+ */
+public class PackException extends Exception {
+
+    public PackException(String message) {
+        super(message);
+    }
+
+    public PackException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiConsumer.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiConsumer.java
new file mode 100644
index 0000000000000000000000000000000000000000..467266f080c12d555db4b10025b4b5162ad5ed41
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiConsumer.java
@@ -0,0 +1,14 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+import java.util.function.BiConsumer;
+
+/**
+ * A {@link BiConsumer} that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingBiConsumer<T, U> {
+
+    void accept(T t, U u) throws PackException;
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiFunction.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiFunction.java
new file mode 100644
index 0000000000000000000000000000000000000000..f4d1885c160a65611ce70f2a4636f1e4cafda29f
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingBiFunction.java
@@ -0,0 +1,14 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+import java.util.function.BiFunction;
+
+/**
+ * A {@link BiFunction} that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingBiFunction<T, U, R> {
+
+    R apply(T t, U u) throws PackException;
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingConsumer.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingConsumer.java
new file mode 100644
index 0000000000000000000000000000000000000000..d4a74f148f332d29c1e169d1edf167c8a0570313
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingConsumer.java
@@ -0,0 +1,14 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+import java.util.function.Consumer;
+
+/**
+ * A {@link Consumer} that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingConsumer<T> {
+
+    void accept(T t) throws PackException;
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingFunction.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingFunction.java
new file mode 100644
index 0000000000000000000000000000000000000000..d903cb1cb0b553d47995f19d48df100889e53079
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingFunction.java
@@ -0,0 +1,14 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+import java.util.function.Function;
+
+/**
+ * A {@link Function} that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingFunction<T, R> {
+
+    R apply(T t) throws PackException;
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingQuadFunction.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingQuadFunction.java
new file mode 100644
index 0000000000000000000000000000000000000000..42fd1463bdab6408ebfd3af2b10a68c58ffb6f7f
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingQuadFunction.java
@@ -0,0 +1,12 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+/**
+ * A function with four arguments, that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingQuadFunction<T, U, V, W, R> {
+
+    R apply(T t, U u, V v, W w) throws PackException;
+
+}
diff --git a/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingTriConsumer.java b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingTriConsumer.java
new file mode 100644
index 0000000000000000000000000000000000000000..fca7ed28c68ea6be7b2b177748f3fdd75a3e9799
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/pack/exception/PackExceptionThrowingTriConsumer.java
@@ -0,0 +1,12 @@
+// Fiddle - read packs - exception class
+
+package org.fiddlemc.fiddle.pack.exception;
+
+/**
+ * A consumer with three arguments, that may throw a {@link PackException}.
+ */
+public interface PackExceptionThrowingTriConsumer<T, U, V> {
+
+    void accept(T t, U u, V v) throws PackException;
+
+}
